### PR TITLE
adding root_external_ids on asset methods

### LINKED ISSUE #531: Improve API documentation for "List assets"
In https://cognite-sdk-python.readthedocs-hosted.com/en/latest/cognite.html#list-assets:

I could not find any example for assets.list() with using arguemnt "root_ids" which I was told could include either {"id": 123} or {"externalId": "abc"}. 

It is not obvious what Dict in (List[Dict[str, Any]])  in the documentation could and should include and the documentation would benefit a lot from having those types of examples.

Thank you!

### PR DESCRIPTION
**Pull request by sanderland**
should also fix #531 so tagging @lynochka to check as well.

### CODE REVIEW

**Comment by sanderland** [comment]
This fails on
test_list_and_iter_signatures_same_as_filter_signature
@erlendvollset can we disable this check?

File: tests/tests_unit/test_api/test_assets.py [hunk 1]
```diff
@@ -109,6 +109,43 @@ def test_list(self, mock_assets_response):
         assert "bla" == jsgz_load(mock_assets_response.calls[0].request.body)["filter"]["name"]
         assert mock_assets_response.calls[0].response.json()["items"] == res.dump(camel_case=True)
 
+    def test_list_root(self, mock_assets_response):
+        res = ASSETS_API.list(root_ids=[{"id": 1}, {"externalId": "abc"}], limit=10)
```
Comment (erlendvollset) [inline tests/tests_unit/test_api/test_assets.py]: res not used

File: tests/tests_unit/test_api/test_assets.py [hunk 2]
```diff
@@ -109,6 +109,43 @@ def test_list(self, mock_assets_response):
...
+        calls = mock_assets_response.calls
+        assert 1 == len(calls)
+        assert {"cursor": None, "limit": 10, "filter": {"rootIds": [{"id": 1}, {"externalId": "abc"}]}} == jsgz_load(
+            calls[0].request.body
+        )
+
+    def test_call_root(self, mock_assets_response):
+        for a in ASSETS_API(root_ids=[{"id": 1}, {"externalId": "abc"}], limit=10):
```
Comment (erlendvollset) [inline tests/tests_unit/test_api/test_assets.py]: ASSETS_API.__call__(root_ids=[{"id": 1}, {"externalId": "abc"}], limit=10)

File: tests/tests_unit/test_api/test_assets.py [hunk 3]
```diff
@@ -109,6 +109,43 @@ def test_list(self, mock_assets_response):
...
+            pass
+        calls = mock_assets_response.calls
+        assert 1 == len(calls)
+        assert {"cursor": None, "limit": 10, "filter": {"rootIds": [{"id": 1}, {"externalId": "abc"}]}} == jsgz_load(
+            calls[0].request.body
+        )
+
+    def test_list_root_ids_list(self, mock_assets_response):
+        res = ASSETS_API.list(root_ids=[1, 2], limit=10)
```
Comment (erlendvollset) [inline tests/tests_unit/test_api/test_assets.py]: res not used

File: tests/tests_unit/test_api/test_assets.py [hunk 4]
```diff
@@ -109,6 +109,43 @@ def test_list(self, mock_assets_response):
...
+        calls = mock_assets_response.calls
+        assert 1 == len(calls)
+        assert {"cursor": None, "limit": 10, "filter": {"rootIds": [{"id": 1}, {"id": 2}]}} == jsgz_load(
+            calls[0].request.body
+        )
+
+    def test_list_root_extids_list(self, mock_assets_response):
+        res = ASSETS_API.list(root_external_ids=["1", "2"], limit=10)
```
Comment (erlendvollset) [inline tests/tests_unit/test_api/test_assets.py]: res not used

File: cognite/client/_api/assets.py [hunk 5]
```diff
@@ -48,6 +50,15 @@ def __call__(
             Union[Asset, AssetList]: yields Asset one by one if chunk is not specified, else AssetList objects.
         """
 
+        if root_ids and isinstance(root_ids[0], dict):
+            if root_external_ids:
+                raise ValueError(
+                    "When using object syntax for root_ids, passing root_external_ids is not supported. Using separate lists for root_ids, root_external_ids is recommended."
+                )
+        elif root_ids or root_external_ids:
+            root_ids = [{"id": id} for id in root_ids or []] + [{"externalId": id} for id in root_external_ids or []]
```
Comment (erlendvollset) [inline cognite/client/_api/assets.py]: can use `self._process_ids(root_ids, root_external_ids, wrap_ids=True)`

File: cognite/client/_api/assets.py [hunk 6]
```diff
@@ -48,6 +50,15 @@ def __call__(
...
+            print(root_ids)
```
Comment (erlendvollset) [inline cognite/client/_api/assets.py]: Remove print

File: cognite/client/_api/assets.py [hunk 7]
```diff
@@ -35,7 +36,8 @@ def __call__(
             chunk_size (int, optional): Number of assets to return in each chunk. Defaults to yielding one asset a time.
             name (str): Name of asset. Often referred to as tag.
             parent_ids (List[int]): No description.
-            root_ids (List[Dict[str, Any]]): List of root ids/external ids to filter on.
+            root_ids (Union[List[int], List[Dict[str, Any]], optional): List of root ids ids to filter on. Also accepts a list of mixed id/external id objects of the form [{"id":1},{"externalId":"str"}] for backwards compatibility.
```
Comment (erlendvollset) [inline cognite/client/_api/assets.py]: Don't think we should document the fact that you can pass a list of objects. We will want to deprecate the behaviour at some point, so we don't want people to actually use it, only keep support for backwards compatibility. So remove the type hints aswell.

File: cognite/client/_api/assets.py [hunk 8]
```diff
@@ -48,6 +50,15 @@ def __call__(
             Union[Asset, AssetList]: yields Asset one by one if chunk is not specified, else AssetList objects.
         """
 
+        if root_ids and isinstance(root_ids[0], dict):
+            if root_external_ids:
+                raise ValueError(
```
Comment (erlendvollset) [inline cognite/client/_api/assets.py]: Feel like this validation is redundant if we don't document the use of objects for `root_ids`.

File: tests/tests_unit/test_api/test_assets.py (hunk 3 above)
Comment (sanderland) [reply tests/tests_unit/test_api/test_assets.py]: fixed

File: tests/tests_unit/test_api/test_assets.py (hunk 1 above)
Comment (sanderland) [reply tests/tests_unit/test_api/test_assets.py]: fixed

File: tests/tests_unit/test_api/test_assets.py (hunk 4 above)
Comment (sanderland) [reply tests/tests_unit/test_api/test_assets.py]: fixed

File: cognite/client/_api/assets.py (hunk 6 above)
Comment (sanderland) [reply cognite/client/_api/assets.py]: :scream:

File: cognite/client/_api/assets.py (hunk 5 above)
Comment (sanderland) [reply cognite/client/_api/assets.py]: nice!

File: cognite/client/_api/assets.py (hunk 7 above)
Comment (sanderland) [reply cognite/client/_api/assets.py]: ok

File: cognite/client/_api/assets.py (hunk 8 above)
Comment (sanderland) [reply cognite/client/_api/assets.py]: fair enough

File: tests/tests_unit/test_api/test_assets.py (hunk 2 above)
Comment (sanderland) [reply tests/tests_unit/test_api/test_assets.py]: you mean an explicit __call__ ? don't think it matters, but ok

Comment (erlendvollset) [reply tests/tests_unit/test_api/test_assets.py]: no, sorry. just meant that iterating over it is redundant.

Comment (sanderland) [reply tests/tests_unit/test_api/test_assets.py]: it doesn't seem to execute/call when you don't iterate over anything. iterator magic i suppose

**Comment by sanderland** [comment]
the openapi generator tests are doing something reaaaalllyyy weird, adding aggregates to parameters

File: tests/tests_unit/test_api/test_assets.py (hunk 2 above)
Comment (sanderland) [reply tests/tests_unit/test_api/test_assets.py]: changed it to list() which works too

File: cognite/client/_api/assets.py [hunk 9]
```diff
@@ -35,7 +36,8 @@ def __call__(
             chunk_size (int, optional): Number of assets to return in each chunk. Defaults to yielding one asset a time.
             name (str): Name of asset. Often referred to as tag.
             parent_ids (List[int]): No description.
-            root_ids (List[Dict[str, Any]]): List of root ids/external ids to filter on.
+            root_ids (Union[List[int], List[Dict[str, Any]], optional): List of root ids ids to filter on.
```
Comment (erlendvollset) [inline cognite/client/_api/assets.py]: Remove the List[Dict[str, Any]] type hint

File: cognite/client/_api/assets.py [hunk 10]
```diff
@@ -144,7 +150,8 @@ def list(
         Args:
             name (str): Name of asset. Often referred to as tag.
             parent_ids (List[int]): List of parent ids to filter on.
-            root_ids (List[Dict[str, Any]]): List of root ids/root external ids to filter on.
+            root_ids (Union[List[int], List[Dict[str, Any]], optional): List of root ids ids to filter on.
```
Comment (erlendvollset) [inline cognite/client/_api/assets.py]: Remove the List[Dict[str, Any]] type hint

**Comment by sanderland** [comment]
the openapi generator tests are getting quite annoying now that starbase is actually updating their docs 0_0

**Pull request merged**